enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,213 @@
+1. microshift
+
+Nothing would change in this case, installing microshift, which becomnes a virtual
+package, installs the necessary dependencies:
+* microshift-core
+* microshift-networking-ovn
+* microshift-storage-topolvm
+* microshift-routes
+
+as well as any other dependencies.
+
+An application developer only needs advanced networking, but no persistent volumes,
+and no routes/ingress.  Creates an image that contains:
+1. microshift-core
+2. microshift-networking-ovn
+
+
+An application developer wants the bare minimum set of components creates an image
+with:
+1. microshift-core
+
+An application developer wanting bare minimum networking but persistent volumes
+creates an image with:
+1. microshift-core
+2. microshift-storage-topolvm
+
+### Goals
+* Allow networking, storage and router core components to be optionally
+  installed.
+
+* As we refactor and re-use the kustomize package used for loading user workloads,
+  add support for loading yamls from multiple directories, i.e.:
+  * `/usr/lib/microshift/manifests.d/user-app-1/`
+  * `/usr/lib/microshift/manifests.d/user-app-2/`
+  * `/etc/microshift/manifests.d/user-app-3/`
+  * `...`
+
+* Split the RPM packages to resemble those optional components.
+
+* Preserve previous behavior in terms of RPM package installation.
+
+### Non-Goals
+
+* Making DNS component optional.
+* Making the certificate management component optional.
+
+## Proposal
+
+To allow modularity of the described core components, the following changes
+would be introduced:
+
+Update the service manager to allow core components as kustomization yamls,
+in the same way we already support for user application manifests. i.e.:
+
+`/usr/lib/microshift/components.d/networking-ovn/kustomize.yaml`
+`/usr/lib/microshift/components.d/storage-topolvm/kustomize.yaml`
+`/usr/lib/microshift/components.d/openshift-router/kustomize.yaml`
+
+  MicroShift component manager would scan the `/usr/lib/microshift/components.d/`
+  directory and extract metadata from the yamls to analyze dependencies between
+  core components at boot. Then start the right componments in the right order.
+
+* The kustomize component loader would be refactored so it can be used for this
+  purpose. We should be able to pre-process kustomize yamls to insert template
+  variables as we do today for core assets.
+
+The following changes would be introduced to packaging:
+
+* Convert `microshift` into a virtual package that depends on `microshift-core`,
+  `microshift-networking-ovn`, `microshift-storage-topolvm`, `microshift-openshift-router`
+  to preserve the current behavior.
+
+* Convert the `microshift-networking` package would be made a virtual package which
+  depends on `microshift-networking-ovn` to preserver current behavior.
+
+* The `microshift-networking-ovn` package would include the necessary
+  `/usr/lib/microshift/components.d/networking-ovn/` yamls, including crds.
+
+* The `microshift-storage-topolvm` package would include the necessary
```

Comment:
Should this be symmetric with the ovn package dependencies? I mean, should we create microshift-storage virtual package that depends on microshift-storage-topolvm?

Reply:
It did it mostly to ensure backwards compatibility to blueprints, because today we have microshift-networking as a package, but we do not have microshift-storage.

I suspect we would prefer users to start consuming microshift-networking-ovn / microshift-storage-topolvm to be more explicit. For example, if any of those plugins are removed in the future, the consumers get an error and they know they need to take action and understand what the changes were, otherwise, the plugin could change under their feet and create unexpected behavior on their product.

On the other hand, what you suggest would make it more symmetrical to consume. May be we should create the virtual package, but at the same time, mark and document as deprecated if what I say above makes sense.

Reply:
We are in dev preview stage, so we're not supposed to maintain any backward compatibility. Let's make it so our story 4.13 onwards is consistent.